subsys/disk/disk_access_stm32_sdmmc.c

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (c) 2020 Amarula Solutions.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_sdmmc
+
+#include <devicetree.h>
+#include <disk/disk_access.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <drivers/gpio.h>
+#include <logging/log.h>
+#include <soc.h>
+
+LOG_MODULE_REGISTER(stm32_sdmmc);
+
+struct stm32_sdmmc_priv {
+	SD_HandleTypeDef hsd;
+	int status;
+	struct k_work work;
+	struct gpio_callback cd_cb;
+	struct {
+		const char *name;
+		struct device *port;
+		int pin;
+		int flags;
+	} cd;
+	struct {
+		const char *name;
+		struct device *port;
+		int pin;
+		int flags;
+	} pe;
+	struct stm32_pclken pclken;
+};
+
+static int stm32_sdmmc_clock_enable(struct stm32_sdmmc_priv *priv)
+{
+	struct device *clock;
+
+#if CONFIG_SOC_SERIES_STM32L4X
+	LL_RCC_PLLSAI1_Disable();
+	/* Configure PLLSA11 to enable 48M domain */
+	LL_RCC_PLLSAI1_ConfigDomain_48M(LL_RCC_PLLSOURCE_HSI,
+					LL_RCC_PLLM_DIV_1,
+					8, LL_RCC_PLLSAI1Q_DIV_8);
+
+	/* Enable PLLSA1 */
+	LL_RCC_PLLSAI1_Enable();
+
+	/*  Enable PLLSAI1 output mapped on 48MHz domain clock */
+	LL_RCC_PLLSAI1_EnableDomain_48M();
+
+	/* Wait for PLLSA1 ready flag */
+	while (LL_RCC_PLLSAI1_IsReady() != 1)
+		;
+
+	LL_RCC_SetSDMMCClockSource(LL_RCC_SDMMC1_CLKSOURCE_PLLSAI1);
+#endif
+
+	clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	if (!clock) {
+		return -ENODEV;
+	}
+
+	/* Enable the APB clock for stm32_sdmmc */
+	return clock_control_on(clock, (clock_control_subsys_t *)&priv->pclken);
+}
+
+static int stm32_sdmmc_clock_disable(struct stm32_sdmmc_priv *priv)
+{
+	struct device *clock;
+
+	clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	if (!clock) {
+		return -ENODEV;
+	}
+
+	return clock_control_off(clock,
+				 (clock_control_subsys_t *)&priv->pclken);
+}
+
+static int stm32_sdmmc_access_init(struct disk_info *disk)
+{
+	struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->driver_data;
+	int err;
+
+	if (priv->status == DISK_STATUS_OK) {
+		return 0;
+	}
+
+	if (priv->status == DISK_STATUS_NOMEDIA) {
+		return -ENODEV;
+	}
+
+	err = stm32_sdmmc_clock_enable(priv);
+	if (err) {
+		LOG_ERR("failed to init clocks");
+		return err;
+	}
+
+	err = HAL_SD_Init(&priv->hsd);
+	if (err != HAL_OK) {
+		LOG_ERR("failed to init stm32_sdmmc");
+		return -EIO;
+	}
+
+	priv->status = DISK_STATUS_OK;
+	return 0;
+}
+
+static void stm32_sdmmc_access_deinit(struct stm32_sdmmc_priv *priv)
+{
+	HAL_SD_DeInit(&priv->hsd);
+	stm32_sdmmc_clock_disable(priv);
+}
+
+static int stm32_sdmmc_access_status(struct disk_info *disk)
+{
+	struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->driver_data;
+
+	LOG_DBG("%s", __func__);
```

Comment:
Can you check that LOG_DBG is not already returning function name ?

Reply:
I think it does. Should I just remove all the DBG messages?